File: greedy/435.Non-overlapping_Intervals.py
# Array, Greedy, Sorting
# Recall 253. Meeting Rooms II
# Similar to 452. Minimum Number of Arrows to Burst Balloons

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        # # LIS ver
        # # O(n^2), will exceed time limits for some tests!
        # if len(intervals) <= 1:
        #     return 0

        # intervals.sort()
        
        # dp = [1 for _ in range(len(intervals))]
        # for i in range(1, len(intervals)):
        #     for j in range(i):
        #         if intervals[j][1] <= intervals[i][0]:
        #             dp[i] = max(dp[i], dp[j] + 1)

        # return len(intervals) - max(dp)

        # Greedy ver
        if len(intervals) <= 1:
            return 0

        non_overlapping_count = 1
        intervals.sort(key=lambda x : x[1])
        prev_end = intervals[0][1]
        for i in range(1, len(intervals)):
            curr_start = intervals[i][0]
            if curr_start >= prev_end:
                non_overlapping_count += 1
                prev_end = intervals[i][1]

        return len(intervals) - non_overlapping_count

File: greedy/435.Non-overlapping_Intervals.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {

        if (intervals.size() < 2)
            return 0;
        
        // 1) Use LIS, O(N^2), might exceed time limits for some tests
        // Solution skipped here

        // 2) Greedy: consider to retain interval with sooner/earlier end time
        sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b){
            return a[1] < b[1];
        });
        int non_overlapping_count = 1;
        int prev_end_time = intervals[0][1];
        for (int i = 1; i < intervals.size(); ++i)
        {
            if (intervals[i][0] >= prev_end_time)
            {
                prev_end_time = intervals[i][1];
                ++non_overlapping_count;
            }
        }
        return intervals.size() - non_overlapping_count;
    }
};
